sdcflows/transform.py

```diff
@@ -72,151 +274,243 @@ def fit(self, spatialimage):
         else:
-            # Map coordinates from reference to time-step
-            self.xfm.reference = spatialimage
-            hmc_xyz = self.xfm.map(self.xfm.reference.ndcoords.T)
-            # Convert from RAS to voxel coordinates
-            voxcoords = (
-                np.linalg.inv(self.xfm.reference.affine)
-                @ _as_homogeneous(np.vstack(hmc_xyz), dim=self.xfm.reference.ndim).T
-            )[:3, ...]
-
-        # fmap * ro_time is the voxel-shift map (VSM)
-        # The VSM is just the displacements field given in index coordinates
-        # voxcoords is the deformation field, i.e., the target position of each voxel
-        voxcoords[pe_axis, ...] += fmap.reshape(-1) * ro_time
-
-        # Prepare data
-        data = np.squeeze(np.asanyarray(spatialimage.dataobj))
-        output_dtype = output_dtype or data.dtype
+            # Generate warp field (before ensuring positive cosines)
+            self.fit(moving, fmap2data_xfm=fmap2data_xfm, approx=approx)
+
+        # Squeeze non-spatial dimensions
+        newshape = moving.shape[:3] + tuple(dim for dim in moving.shape[3:] if dim > 1)
+        data = np.reshape(moving.dataobj, newshape)
+        ndim = min(data.ndim, 3)
+        n_volumes = data.shape[3] if data.ndim == 4 else 1
+        output_dtype = output_dtype or moving.header.get_data_dtype()
+
+        # Prepare input parameters
+        if isinstance(pe_dir, str):
+            pe_dir = [pe_dir]
+
+        if isinstance(ro_time, float):
+            ro_time = [ro_time]
+
+        if n_volumes > 1 and len(pe_dir) == 1:
+            pe_dir *= n_volumes
+
+        if n_volumes > 1 and len(ro_time) == 1:
+            ro_time *= n_volumes
+
+        pe_info = []
+        for volid in range(n_volumes):
+            pe_axis = "ijk".index(pe_dir[volid][0])
+            axis_flip = axcodes[pe_axis] in ("LPI")
+            pe_flip = pe_dir[volid].endswith("-")
+
+            pe_info.append((
+                pe_axis,
+                # Displacements are reversed if either is true (after ensuring positive cosines)
+                -ro_time[volid] if (axis_flip ^ pe_flip) else ro_time[volid]
+            ))
+
+        # Reference image's voxel coordinates (in voxel units)
+        voxcoords = nt.linear.Affine(
+            reference=moving
+        ).reference.ndindex.reshape((ndim, *data.shape[:ndim])).astype("float32")
+
+        # Convert head-motion transforms to voxel-to-voxel:
+        if xfms is not None:
+            if len(xfms) != n_volumes:
+                raise RuntimeError(
+                    f"Number of head-motion estimates ({len(xfms)}) does not match the "
+                    f"number of volumes ({n_volumes})"
+                )
+            vox2ras = moving.affine.copy()
+            ras2vox = np.linalg.inv(vox2ras)
+            xfms = [ras2vox @ xfm @ vox2ras for xfm in xfms]
 
         # Resample
-        resampled = ndi.map_coordinates(
+        resampled = asyncio.run(unwarp_parallel(
```

Comment:
I'm curious to see how performance is affected by switching this to work within an async framework - just noting that this will limit the use of this class/function in cases of an already running event loop (without hacks like nest_asyncio)

Reply:
I'd say not a lot better than a thread executor, but much better than a processor executor or joblib

Reply:
Here the main bottleneck is memory.